pkg/controller/nodeipam/ipam/clustercidrconfig_controller.go

```diff
@@ -0,0 +1,484 @@
+	ipv6MaxCIDRMask            = 128
+	// maxRetries is the number of times a deployment will be retried before it is dropped out of the queue.
+	// With the current rate-limiter in use (5ms*2^(maxRetries-1)) the following numbers represent the times
+	// a deployment is going to be requeued:
+	//
+	// 5ms, 10ms, 20ms, 40ms, 80ms, 160ms, 320ms, 640ms, 1.3s, 2.6s, 5.1s, 10.2s, 20.4s, 41s, 82s
+	maxRetries                   = 15
+	defaultClusterCIDRConfigName = "default-cluster-cidr-config"
+)
+
+// ClusterCIDRConfigController reconciles ClusterCIDRConfig objects.
+type ClusterCIDRConfigController struct {
+	kubeClient clientset.Interface
+	// Recorder to record K8S events created by the controller.
+	recorder record.EventRecorder
+	lock     *sync.Mutex
+
+	clusterCIDRConfigLister networkinglisters.ClusterCIDRConfigLister
+	clusterCIDRConfigSynced cache.InformerSynced
+
+	queue workqueue.RateLimitingInterface
+
+	// CIDRMap keeps a mapping between the label and the associated cluster CIDRs.
+	CIDRMap map[string][]*cidrset.ClusterCIDR
+}
+
+func NewClusterCIDRConfigController(
+	clusterCIDRConfigInformer networkinginformers.ClusterCIDRConfigInformer,
+	kubeClient clientset.Interface,
+	allocatorParams CIDRAllocatorParams) (*ClusterCIDRConfigController, error) {
+	if kubeClient == nil {
+		klog.Fatalf("kubeClient is nil when starting Controller")
+	}
+
+	eventBroadcaster := record.NewBroadcaster()
+	recorder := eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{
+		Component: "clusterCIDRConfigController",
+	})
+	eventBroadcaster.StartStructuredLogging(0)
+
+	klog.Infof("Sending events to api server.")
+	eventBroadcaster.StartRecordingToSink(
+		&v1core.EventSinkImpl{
+			Interface: kubeClient.CoreV1().Events(""),
+		})
+
+	if kubeClient.CoreV1().RESTClient().GetRateLimiter() != nil {
+		ratelimiter.RegisterMetricAndTrackRateLimiterUsage("cluster_cidr_config_controller", kubeClient.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	cccController := &ClusterCIDRConfigController{
+		kubeClient: kubeClient,
+		CIDRMap:    allocatorParams.CIDRMap,
+		recorder:   recorder,
+		lock:       allocatorParams.CIDRLock,
+		queue:      workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "cluster_cidr_config_controller"),
+	}
+
+	existingConfigList, err := listClusterCIDRConfigs(kubeClient)
+	if err != nil {
+		return nil, err
+	}
+
+	createDefaultClusterCIDRConfig(existingConfigList, allocatorParams)
+
+	// Regenerate the CIDRMaps from the existing ClusterCIDRConfigs.
+	for _, clusterCIDRConfig := range existingConfigList.Items {
+		klog.Infof("Creating ClusterCIDRConfig: %s", clusterCIDRConfig.Name)
+		// Create an event for invalid ClusterCIDRConfigs, do not crash on failures.
+		if err := cccController.reconcileBootstrap(&clusterCIDRConfig); err != nil {
+			klog.Errorf("Error while creating ClusterCIDRConfig: %v", err)
+			cccController.recorder.Event(&clusterCIDRConfig, "Warning", "InvalidClusterCIDRConfig", err.Error())
+		}
+	}
```

Comment:
I see, so when we do `NewClusterCIDRConfigController` we need the apiserver to work ... I don't know if these bootstrap step belong to New() or to Run() or if we should have some PrepareToRun() method , @thockin

Reply:
I think these reconcile* can go after 
```
	if !cache.WaitForNamedCacheSync("cluster_cidr_config_controller", stopCh, r.clusterCIDRConfigSynced) {
		return
	}
```

and before starting the workers, that guarantees that we are in sync with current state and we didn't start processing objects

Reply:
also you can save the `listClusterCIDRConfigs` and iterate over the cache of the informer instead